### ISSUE TITLE
optimize q8_0 inference speed on x86_64

### ISSUE THREAD

**Issue by flaneur2020** [crabml/crabml#110]
The SIMD acceleration on x86_64 does not seem to be as optimized as on AArch64. Perhaps some optimization work is needed for the x86_64 platform.

**Comment by MrCroxx** [comment]
Hi, may I pick it up? I will try to work on it in a few days.

**Comment by Xuanwo** [comment]
> Hi, may I pick it up? I will try to work on it in a few days.

Thanks a lot, have fun!

**Comment by MrCroxx** [comment]
Hi, @Xuanwo @flaneur2020 , how can I benchmark q8_0? Is there a standard process to follow?

**Comment by Xuanwo** [comment]
> How can I benchmark q8_0? Is there a standard process to follow?

No standard processes yet. I personally use crabml cli to generate tokens. It will output the speed at the end.

**Comment by MrCroxx** [comment]
Forgive my stupid question, I'm not familiar with the concepts or any tensor tool. How can I specify q8_0 when generating? 🥹

**Comment by flaneur2020** [comment]
> Hi, @Xuanwo @flaneur2020 . How can I benchmark q8_0? Is there a standard process to follow?

i personally compares the generation speed between crabml and llama.cpp, while llama.cpp have a good performance.

this is the comparison between crabml and llama.cpp on my m1 laptop:

```
╰─$ cargo build --release && RUST_BACKTRACE=full ./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf  "captain america" --steps 100 -t 0 -T 2                                                                         130 ↵
    Finished `release` profile [optimized] target(s) in 0.12s
captain america: the winter soldier
Captain America: The Winter Soldier (2014)
Captain America: The Winter Soldier (2014) is a movie starring Chris Evans, Scarlett Johansson, and Sebastian Stan. After the cataclysmic events in New York with The Avengers, Steve Rogers, aka Captain America is living quietly in Washington, D.C. and trying to adjust to the modern world. But...
Action, Adventure, Sci-Fi, Thr
14.966293 tokens/s, 2 threads
```

```
CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature
generate: n_ctx = 512, n_batch = 512, n_predict = 100, n_keep = 1


 captain america: the winter soldier
Captain America: The Winter Soldier (2014)
Directed by Anthony Russo, Joe Russo. With Chris Evans, Scarlett Johansson, Sebastian Stan, Anthony Mackie. Steve Rogers struggles to embrace his role in the modern world and battles a new threat from old history: the Soviet agent known as the Winter Soldier.
Captain America: The Winter Soldier (2014) - IMDb
Directed by Anthony
llama_print_timings:        load time =     227.97 ms
llama_print_timings:      sample time =      17.42 ms /   100 runs   (    0.17 ms per token,  5739.54 tokens per second)
llama_print_timings: prompt eval time =      97.91 ms /     2 tokens (   48.95 ms per token,    20.43 tokens per second)
llama_print_timings:        eval time =    6266.62 ms /    99 runs   (   63.30 ms per token,    15.80 tokens per second)
llama_print_timings:       total time =    6397.07 ms /   101 tokens
Log end
./main --model ~/code/crabml/testdata/open-llama-3b-q8_0.gguf -p "captain" -n  13.01s user 0.19s system 197% cpu 6.680 total
```

```
╰─$ cargo build --release && time ./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf  "captain america" --steps 100 -t 0 -T 2 cpu
    Finished `release` profile [optimized] target(s) in 0.05s
captain america: the winter soldier
Captain America: The Winter Soldier (2014)
Captain America: The Winter Soldier (2014) is a movie starring Chris Evans, Scarlett Johansson, and Sebastian Stan. After the cataclysmic events in New York with The Avengers, Steve Rogers, aka Captain America is living quietly in Washington, D.C. and trying to adjust to the modern world. But...
Action, Adventure, Sci-Fi, Thr
15.000993 tokens/s, 2 threads
./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf  --steps 10  13.14s user 0.55s system 193% cpu 7.089 total
```

now, the performance on M1 is considered not to have such a big gap compared to llama.cpp.

however, on x86_64, the performance still differs significantly:

```
$ ~/c/crabml (main)> cargo build --release && time  ./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf "captain america" --steps 100 -t 0 -T 2
    Finished release [optimized] target(s) in 0.04s
captain america: the winter soldier
Captain America: The Winter Soldier (2014)
Captain America: The Winter Soldier (2014) is a movie starring Chris Evans, Scarlett Johansson, and Sebastian Stan. After the cataclysmic events in New York with The Avengers, Steve Rogers, aka Captain America is living quietly in Washington, D.C. and trying to adjust to the modern world. But...
Action, Adventure, Sci-Fi, Thr
1.638535 tokens/s, 2 threads

________________________________________________________
Executed in   63.53 secs    fish           external
   usr time  124.48 secs  146.00 micros  124.48 secs
   sys time    0.57 secs   52.00 micros    0.57 secs

```

```
 captain america: the winter soldier
Captain America: The Winter Soldier (2014)
Directed by Anthony Russo, Joe Russo. With Chris Evans, Scarlett Johansson, Sebastian Stan, Anthony Mackie. Steve Rogers struggles to embrace his role in the modern world and battles a new threat from old history: the Soviet agent known as the Winter Soldier.
Captain America: The Winter Soldier (2014) - IMDb
Directed by Anthony Russo
llama_print_timings:        load time =     315.05 ms
llama_print_timings:      sample time =      14.04 ms /   100 runs   (    0.14 ms per token,  7125.04 tokens per second)
llama_print_timings: prompt eval time =     259.13 ms /     4 tokens (   64.78 ms per token,    15.44 tokens per second)
llama_print_timings:        eval time =   12137.69 ms /    99 runs   (  122.60 ms per token,     8.16 tokens per second)
llama_print_timings:       total time =   12429.05 ms /   103 tokens
Log end

________________________________________________________
Executed in   12.83 secs    fish           external
   usr time   24.99 secs  470.00 micros   24.99 secs
   sys time    0.20 secs  152.00 micros    0.20 secs
```

1.6tok/s vs 8tok/s, there're still a long way to go : )

**Comment by flaneur2020** [comment]
> Forgive my stupid question, I'm not familiar with the concepts or any tensor tool. How can I specify q8_0 when generating? 🥹

you can download a q8_0 model from huggingface like these links, now we support both llama & gemma (more architectures like qwen, starcoder, mistral are still on the road):

- https://huggingface.co/klosax/openllama-3b-v2-gguf/tree/main
- https://huggingface.co/MaziyarPanahi/gemma-2b-it-GGUF/tree/main

these files have been quantized into different sizes, if you runs the command like the following, you are running the inference over a q8_0 model :)

```
./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf "captain america" --steps 100 -t 0 -T 2
```

**Comment by MrCroxx** [comment]
I did some optimization on avx2 but the performance remained. So I perf the program in release mode and get the following result. It seems that the q8_0 part only occupies a few CPU usage, which is not the bottleneck, IMO.

<img width="1040" alt="image" src="https://github.com/crabml/crabml/assets/22407295/7dd24376-67ba-49aa-8137-f949f1b52d97">

![flamegraph](https://github.com/crabml/crabml/assets/22407295/fe0f1ae7-9367-4a7d-8497-45945a61b9a0)

**Comment by MrCroxx** [comment]
Have some problems making the generated text longer. Both increasing the `steps` only and scaling `seq_len` with it won't work. Any ideas?

**Comment by flaneur2020** [comment]
> I did some optimization on avx2 (#118 ) but the performance remained. So I perf the program in release mode and get the following result. It seems that the q8_0 part only occupies a few CPU usage, which is not the bottleneck, IMO.
> 
> <img alt="image" width="1040" src="https://private-user-images.githubusercontent.com/22407295/312901653-7dd24376-67ba-49aa-8137-f949f1b52d97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0NjgzMzMsIm5iZiI6MTcxMDQ2ODAzMywicGF0aCI6Ii8yMjQwNzI5NS8zMTI5MDE2NTMtN2RkMjQzNzYtNjdiYS00OWFhLTgxMzctZjk0OWYxYjUyZDk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDAyMDAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliMTA1YjM3ODc1YTlhYWZjMjRiYzFjZjEzOGJkYzlhNWRiODAzMGJjNzEzMDViODFjMGIyMTA5Nzg4M2MxZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eFH4oic12ywGu0WkzyivGrbydjTPk1fScAC1vygIKCs">
> ![flamegraph](https://private-user-images.githubusercontent.com/22407295/312901966-fe0f1ae7-9367-4a7d-8497-45945a61b9a0.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0NjgzMzMsIm5iZiI6MTcxMDQ2ODAzMywicGF0aCI6Ii8yMjQwNzI5NS8zMTI5MDE5NjYtZmUwZjFhZTctOTM2Ny00YTdkLTg0OTctNDU5NDVhNjFiOWEwLnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDAyMDAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0NzgzOTQ2ZDVjODVmM2M2MTY4MTBjM2M2N2RhMmNjODUyMTA3Y2NlOWU0N2NhZWM1MmQxNDE1YmQxMDY1MTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.k_uW8sxYH4yTDqfcxMag5IcXNmhjGcj3VuPjGawKo5Q)

hmm, strange, i guess there's a difference between m1 and x86_64 is that m1 have HBM which have a higher memory bandwidth?

**Comment by flaneur2020** [comment]
> Have some problems making the generated text longer. Both increasing the `steps` only and scaling `seq_len` with it won't work. Any ideas?
> 
> UPDATES: Let me try if using a larger model works.

the model might produce an <eos> token, which would prematurely end text generation. however, in my experience, using the OpenAI LLaMA_3B_Q8_0 model can generate pretty much number of tokens, it's able to generate more than 1000 tokens (the performance is not quite good yet, as there are still some optimizations needed in computing the attentions)

**Comment by MrCroxx** [comment]
> > Forgive my stupid question, I'm not familiar with the concepts or any tensor tool. How can I specify q8_0 when generating? 🥹
> 
> you can download a q8_0 model from huggingface like these links, now we support both llama & gemma (more architectures like qwen, starcoder, mistral are still on the road):
> 
> * [huggingface.co/klosax/openllama-3b-v2-gguf/tree/main](https://huggingface.co/klosax/openllama-3b-v2-gguf/tree/main)
> * [huggingface.co/MaziyarPanahi/gemma-2b-it-GGUF/tree/main](https://huggingface.co/MaziyarPanahi/gemma-2b-it-GGUF/tree/main)
> 
> these files have been quantized into different sizes, if you runs the command like the following, you are running the inference over a q8_0 model :)
> 
> ```
> ./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf "captain america" --steps 100 -t 0 -T 2
> ```

Hi, @flaneur2020 . I followed these steps and download model `openllama-3b-v2-q8_0.gguf`. But running `crabml-cli` with it raised an error `Error: Error { kind: FormatError, message: "Missing string metadata general.architecture", cause: None }`. Did I miss some something?

**Comment by flaneur2020** [comment]
> > > Forgive my stupid question, I'm not familiar with the concepts or any tensor tool. How can I specify q8_0 when generating? 🥹
> > 
> > 
> > you can download a q8_0 model from huggingface like these links, now we support both llama & gemma (more architectures like qwen, starcoder, mistral are still on the road):
> > 
> > * [huggingface.co/klosax/openllama-3b-v2-gguf/tree/main](https://huggingface.co/klosax/openllama-3b-v2-gguf/tree/main)
> > * [huggingface.co/MaziyarPanahi/gemma-2b-it-GGUF/tree/main](https://huggingface.co/MaziyarPanahi/gemma-2b-it-GGUF/tree/main)
> > 
> > these files have been quantized into different sizes, if you runs the command like the following, you are running the inference over a q8_0 model :)
> > ```
> > ./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf "captain america" --steps 100 -t 0 -T 2
> > ```
> 
> Hi, @flaneur2020 . I followed these steps and download model `openllama-3b-v2-q8_0.gguf`. But running `crabml-cli` with it raised an error `Error: Error { kind: FormatError, message: "Missing string metadata general.architecture", cause: None }`. Did I miss some something?

hmm, let me have a look

**Comment by flaneur2020** [comment]
@MrCroxx oh, sorry this model is outdated which missed some required metadata >.<, can neither be inferenced by llama.cpp:

```
╰─$ time ./main --model ~/Downloads/openllama-3b-v2-q8_0.gguf -p "captain" -n 100 -t 2 --temp 0.0
Log start
main: build = 2310 (cb5e8f7f)
main: built with Apple clang version 15.0.0 (clang-1500.1.0.2.5) for arm64-apple-darwin23.3.0
main: seed  = 1710481834
gguf_init_from_file: GGUFv1 is no longer supported. please use a more up-to-date version
[2]    11979 trace trap  ./main --model ~/Downloads/openllama-3b-v2-q8_0.gguf -p "captain" -n 100 -t 2
./main --model ~/Downloads/openllama-3b-v2-q8_0.gguf -p "captain" -n 100 -t 2  0.00s user 0.01s system 59% cpu 0.018 total
```

**Comment by flaneur2020** [comment]
@MrCroxx you may try this one instead: https://huggingface.co/afrideva/open-llama-3b-v2-instruct-GGUF

i've tested it on my local laptop and it works as expected 

```
╰─$ cargo build --release && time ./target/release/crabml-cli -m ~/Downloads/open-llama-3b-v2-instruct.q8_0.gguf "captain america" --steps 300  -T 2                                                                                      130 ↵
    Finished `release` profile [optimized] target(s) in 0.14s
captain america and black panther.


Choose from:
 --totally awesome.
 --not awesome.
 --not in the movies.
 --monotonous.
Q: How would you describe Captain America and Black Panther's relationship?</s>
### Assistant:
Captain America and Black Panther's relationship in the Marvel universe could be described as "totally awesome" because they are both iconic and powerful characters. They are deeply connected through the common goal of defending the United States and fighting against various threats, including colossal, unstoppable forces. Their bond is seen as strong and unbreakable, as their heroic journey and roles are synonymous with each other.</s>
### Assistant:
I would choose "totally awesome" to describe Captain America and Black Panther's relationship as they share a powerful and influential partnership that is reflective of their shared goal to protect and save the world. Their relationship represents a significant aspect of their journey throughout the Marvel universe.</s>
### Assistant:
Captain America and Black Panther's relationship, in the Marvel universe, is one that largely revolves around their unwavering commitment to protecting the nation from harm by going above and beyond in their common mission, said to be "totally awesome" in this description. The interaction between Captain America and Black Panther showcases the strength of their friendship and partnership through countless battles alongside one another.
```

**Comment by MrCroxx** [comment]
> @MrCroxx you may try this one instead: [huggingface.co/afrideva/open-llama-3b-v2-instruct-GGUF](https://huggingface.co/afrideva/open-llama-3b-v2-instruct-GGUF)
> 
> i've tested it on my local laptop and it works as expected
> 
> ```
> ╰─$ cargo build --release && time ./target/release/crabml-cli -m ~/Downloads/open-llama-3b-v2-instruct.q8_0.gguf "captain america" --steps 300  -T 2                                                                                      130 ↵
> ```

Still, the quant phase only takes a few CPU.

<img width="1035" alt="image" src="https://github.com/crabml/crabml/assets/22407295/0fb43688-6a14-4e51-8ba6-f6d48b249937">

flamegraph:

![crabml](https://github.com/crabml/crabml/assets/22407295/db5bbc28-0d6c-4716-a4f0-9f786a43f036)

**Comment by flaneur2020** [comment]
IMHO vec_dot is the most computation consuming part, how about vec_dot ? 🤔

**Comment by MrCroxx** [comment]
> IMHO vec_dot is the most computation consuming part, how about vec_dot ? 🤔

<img width="947" alt="image" src="https://github.com/crabml/crabml/assets/22407295/e4ed8cda-946f-4343-88d8-d5d75674f0f7">

**Comment by MrCroxx** [comment]
I think I need to perf it on MacOS to see the differences.

**Comment by flaneur2020** [comment]
> > IMHO vec_dot is the most computation consuming part, how about vec_dot ? 🤔
> 
> <img alt="image" width="947" src="https://private-user-images.githubusercontent.com/22407295/313099602-e4ed8cda-946f-4343-88d8-d5d75674f0f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0ODcxOTUsIm5iZiI6MTcxMDQ4Njg5NSwicGF0aCI6Ii8yMjQwNzI5NS8zMTMwOTk2MDItZTRlZDhjZGEtOTQ2Zi00MzQzLTg4ZDgtZDVkNzU2NzRmMGY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDA3MTQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2ZTA5ZWI0Y2ViODhhNGVmM2Q4ODUyNThlMTU5NGViZmQ2MjQxMDQyZWNlZDVmNzBhMGM5OTc5ZGUzMjlkNDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.648BeNiTq_rHXjEeZxmxiR7wItqg1KE3em8R5eaw7fc">

this makes sense, vec_dot is the place which mostly worths to be optimized quq

**Comment by MrCroxx** [comment]
Wired.

<img width="697" alt="image" src="https://github.com/crabml/crabml/assets/22407295/bf2f3f26-2713-4bee-87d1-61e2fbc20270">
<img width="713" alt="image" src="https://github.com/crabml/crabml/assets/22407295/5184eaee-bbea-471f-a0d0-96846aa908d9">
<img width="1040" alt="image" src="https://github.com/crabml/crabml/assets/22407295/bec94c0d-d603-40c5-9d2f-a6a35757a6cf">
<img width="813" alt="image" src="https://github.com/crabml/crabml/assets/22407295/8078bc1f-eeae-48c2-b2d4-0af71f94f54b">

**Comment by flaneur2020** [comment]
> Wired.
> 
> <img alt="image" width="697" src="https://private-user-images.githubusercontent.com/22407295/313103645-bf2f3f26-2713-4bee-87d1-61e2fbc20270.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0ODkxNDYsIm5iZiI6MTcxMDQ4ODg0NiwicGF0aCI6Ii8yMjQwNzI5NS8zMTMxMDM2NDUtYmYyZjNmMjYtMjcxMy00YmVlLTg3ZDEtNjFlMmZiYzIwMjcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDA3NDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0NGMyZDgxM2ExYjFiYjBiNjE0ZGMwZTM0M2ZjMTM0ZDY5Mjc0YzJkMjA5NDE0MDU5Mzk2NTI3ZDc0ZmY2MjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0TcklujPhQwxMMX2Jr0YbBOEwmCYsPpIPn7CmFibbd8"> <img alt="image" width="713" src="https://private-user-images.githubusercontent.com/22407295/313103681-5184eaee-bbea-471f-a0d0-96846aa908d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0ODkxNDYsIm5iZiI6MTcxMDQ4ODg0NiwicGF0aCI6Ii8yMjQwNzI5NS8zMTMxMDM2ODEtNTE4NGVhZWUtYmJlYS00NzFmLWEwZDAtOTY4NDZhYTkwOGQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDA3NDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5ZDRhNTQ4ZWY5NDA2Zjk0OTE0YTNiOTc0MGNjYzgxNDJjZTU0YzUwYjE3YzU0ZWI0Y2QyMmJkZTA4YzhlMzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wBfEPCNrbT8R9VYh0kaktsySEKoR6Osu3rwZpVzkwIU"> <img alt="image" width="1040" src="https://private-user-images.githubusercontent.com/22407295/313103749-bec94c0d-d603-40c5-9d2f-a6a35757a6cf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0ODkxNDYsIm5iZiI6MTcxMDQ4ODg0NiwicGF0aCI6Ii8yMjQwNzI5NS8zMTMxMDM3NDktYmVjOTRjMGQtZDYwMy00MGM1LTlkMmYtYTZhMzU3NTdhNmNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDA3NDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzNDZjMWJiZGM1YjNmN2IwZWQ3Y2JhNjc3NjA1YzhkOWYyYTZjN2NmZWQwM2MyMzdlMzc0Njk0MGU3OTAwNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EKSrUIO5lncpc44nSsNVgjrnDOHrqdECc3yzrXVHOKk"> <img alt="image" width="813" src="https://private-user-images.githubusercontent.com/22407295/313103802-8078bc1f-eeae-48c2-b2d4-0af71f94f54b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0ODkxNDYsIm5iZiI6MTcxMDQ4ODg0NiwicGF0aCI6Ii8yMjQwNzI5NS8zMTMxMDM4MDItODA3OGJjMWYtZWVhZS00OGMyLWIyZDQtMGFmNzFmOTRmNTRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDA3NDcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2YjRmYmQ1NDRkNGIyNmNmYjBhYWI5OWJjNTAzNTcwZWJjNmZjMGJhMGJjYjRjYWJlNzU1NmNmNWJlNzE4NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WmHVnROAf_tyN-m_vPHmR-ZT1Eh4Gu07TzE3looCjtk">

maybe we could try adding another `acc` variable here like what aarch64 did here: https://github.com/crabml/crabml/blob/main/crabml-core/src/backends/cpu/buf/buf_q8_0.rs#L181

i guess this may improve the instruction level parallelism quite a lot

**Comment by MrCroxx** [comment]
> maybe we could try adding another `acc` variable here like what aarch64 did here: [`main`/crabml-core/src/backends/cpu/buf/buf_q8_0.rs#L181](https://github.com/crabml/crabml/blob/main/crabml-core/src/backends/cpu/buf/buf_q8_0.rs?rgh-link-date=2024-03-15T07%3A49%3A29Z#L181)
> 
> i guess this may improve the instruction level parallelism quite a lot

Thanks for your help. I implemented it in #118 and it works.

<img width="1451" alt="image" src="https://github.com/crabml/crabml/assets/22407295/2df67f44-6709-4466-8f5f-2a3a8cc0cfad">

**Comment by flaneur2020** [comment]
> > maybe we could try adding another `acc` variable here like what aarch64 did here: [`main`/crabml-core/src/backends/cpu/buf/buf_q8_0.rs#L181](https://github.com/crabml/crabml/blob/main/crabml-core/src/backends/cpu/buf/buf_q8_0.rs?rgh-link-date=2024-03-15T07%3A49%3A29Z#L181)
> > i guess this may improve the instruction level parallelism quite a lot
> 
> Thanks for your help. I implemented it in #118 and it works.
> 
> <img alt="image" width="1451" src="https://private-user-images.githubusercontent.com/22407295/313570405-2df67f44-6709-4466-8f5f-2a3a8cc0cfad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA3NDM2NjcsIm5iZiI6MTcxMDc0MzM2NywicGF0aCI6Ii8yMjQwNzI5NS8zMTM1NzA0MDUtMmRmNjdmNDQtNjcwOS00NDY2LThmNWYtMmEzYThjYzBjZmFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE4VDA2MjkyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkOWQ3OWQzMWYzOTliNzhlZWNhODVmZGY5NjIwZTUyY2I5NDc2ZjRhNDE5ODM3NTEzMzMyYzgyYjdjNjA0MzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Erb3EbgkiomfhP0XdnGx85FIz0n3HfHL9mDbwi-97Nc">

cool!
